src/Agent.Worker/StepsRunner.cs

```diff
@@ -205,5 +204,36 @@ public sealed class StepsRunner : AgentService, IStepsRunner
                 Trace.Info($"Current state: job state = '{jobContext.Result}', step failed = {stepFailed}, critical step failed = {criticalStepFailed}");
             }
         }
+
+        private bool ShouldRunStep(IStep step, bool stepFailed, bool criticalStepFailed)
+        {            
+            switch(step.RunMode)
+            {
+                // The server hasn't been upgraded yet. Fallback to old code
+                case RunMode.Undefined:
```

Comment:
RunMode can either be set to:
Default - run when all previous tasks have passed
Always - run always unless a critical step has failed
Falling back to RunMode.Default wouldn't cover the case of the step being marked as RunAlways.

As the agent is common between build and release services and let's say TFS upgrades first along with the agent, data coming from build would be of the latest format but the data coming from RM would still be of the old format. The legacy properties were retained to handle this case